src/http_interface.py

```diff
@@ -0,0 +1,205 @@
+
+        Args:
+            charm: The charm implementing the requirer or provider.
+            relation_name: Name of the integration using the interface.
+        """
+        super().__init__(charm, relation_name)
+
+        observe = self.framework.observe
+        self.charm: CharmBase = charm
+        self.relation_name = relation_name
+
+        observe(charm.on[relation_name].relation_created, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+        observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_departed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    @abc.abstractmethod
+    def _on_relation_joined(self, _: RelationJoinedEvent) -> None:
+        """Abstract method to handle relation-joined event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _on_relation_changed(self, _: RelationChangedEvent) -> None:
+        """Abstract method to handle relation-changed event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _on_relation_broken(self, _: RelationBrokenEvent) -> None:
+        """Abstract method to handle relation-changed event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @property
+    def relations(self) -> list[Relation]:
+        """The list of Relation instances associated with the charm."""
+        relations = self.charm.model.relations.get(self.relation_name, [])
+        return relations
+
+    @property
+    def bind_address(self) -> str:
+        """Get Unit bind address.
+
+        Returns:
+            The unit address, or an empty string if no address found.
+        """
+        if bind := self.model.get_binding("juju-info"):
+            return str(bind.network.bind_address)
+        return ""
+
+
+class HTTPProvider(_IntegrationInterfaceBaseClass):
+    """HTTP interface provider class to be instantiated by the haproxy-operator charm.
+
+    Attrs:
+        on: Custom events that are used to notify the charm using the provider.
+        services: Current services definition parsed from relation data.
+    """
+
+    on = HTTPProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Initialize the HTTPProvider class and parse the relation data.
+
+        Args:
+            charm: The charm instance
+            relation_name: The name of the relation
+        """
+        super().__init__(charm, relation_name)
+        self.services = legacy.generate_service_config(self.get_services_definition())
```

Comment:
There are 2 legacy methods called from this line, which a lot of business logic in them and no error handling in them. I'm wondering if that's a good idea to call this from the constructor

Reply:
Updated to move this code to a method and remove the init method to let the exception flow to the charm as we discussed.